### ISSUE TITLE
Cannot read property 'filter' of undefined

### ISSUE THREAD

**Issue by JulienLeal** [Charca/bootbot#12]
Sometimes when I send message in messenger, I receive this error 

```
TypeError: Cannot read property 'filter' of undefined
    at Conversation.respond (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/Conversation.js:61:53)
    at _conversations.forEach.convo (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:344:22)
    at Array.forEach (native)
    at BootBot._handleConversationResponse (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:341:25)
    at BootBot._handleMessageEvent (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:291:14)
    at entry.messaging.forEach (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:377:20)
    at Array.forEach (native)
    at data.entry.forEach (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:370:27)
    at Array.forEach (native)
    at app.post (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/bootbot/lib/BootBot.js:368:18)
    at Layer.handle [as handle_request] (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/route.js:131:13)
    at Route.dispatch (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/layer.js:95:5)
    at /Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/index.js:277:22
    at Function.process_params (/Users/iurisouz4/Documents/Julien/Projects/facebook_bootbot/node_modules/express/lib/router/index.js:330:12)
```

After that, I receive this error :

```
Error sending message: FetchError: request to https://graph.facebook.com/v2.6/me/messages?access_token=EAAPBAL6ZBQEkBAMIblP8n6wISEjVJVaea4B54zghnwvy87ZAiCE1uexYSt8KipGqBQjzTeCG71hyWoB3wBZBHo33wX9ozH8x1cmgM7VxntkNy01ZABlpKp8Xp0zZC1vgxjIgFBzDQ58EzakeW0IMC58ZAnhGLbI0YKMfuRUKa4g2GUrzAOmlZAL failed, reason: connect ETIMEDOUT 31.13.73.1:443
```

But this second error don't appear everytime.

Actually I'm using ngrok for tunneling it. Do you think that is my problem?  Or is it a bug? Because sometimes I have delay too, both in my project and in your examples.

**Comment by Charca** [comment]
Hi @JulienLeal, it doesn't seem to be related to using ngrok, so it's probably a bug or an unhandled exception. Do you have a code example of the function call that is generating this error?

**Comment by JulienLeal** [comment]
Follow my code
```
'use strict';
const BootBot = require('bootbot');


const bot = new BootBot({
    accessToken: 'xxxxx',
    verifyToken: 'xxxxx',
    appSecret: 'xxxxx'
});
bot.on('message', (payload, chat,data) => {

    chat.conversation((convo) => {
        convo.sendTypingIndicator(1000).then(() => askCampaign(convo));
    });


    const askCampaign = (convo) => {
        convo.ask((convo) => {
            const buttons = [
                { type: 'postback', title: 'A', payload: '1' },
                { type: 'postback', title: 'B', payload: '2' }
            ];
            convo.sendButtonTemplate(`Hello, select the campaign:`, buttons);
        }, (payload, convo, data) => {
            const text = payload.message.text;

        }, [
            {
                event: 'postback:1',
                callback: (payload, convo) => {
                    convo.set('campaign_id', 1);
                    convo.say('You selected A').then(() => askName(convo));
                }
            },
            {
                event: 'postback:2',
                callback: (payload, convo) => {
                    convo.set('campaign_id', 2);
                    convo.say('You selected B').then(() => askName(convo));
                }
            },

        ]);
    };


    const askName = (convo) => {
        convo.ask(`What�s your name?`, (payload, convo, data) => {
            const text = payload.message.text;

            convo.set('name', text);
            if(text!=undefined)
                convo.say(`Your name is ${text}`).then(() => askEmail(convo));
            else
                chat.conversation((convo) => {
                    convo.sendTypingIndicator(1000).then(() => askName(convo));
                });
        });
    };

    const askEmail = (convo) => {
        convo.ask(`What is your email?`, (payload, convo, data) => {
            const text = payload.message.text;
            convo.set('email', text);
            if(text!=undefined)
                convo.say(`Your email is ${text}`).then(() => askGender(convo));
            else
                chat.conversation((convo) => {
                    convo.sendTypingIndicator(1000).then(() => askEmail(convo));
                });
        });
    };

    const askGender = (convo) => {
        convo.ask((convo) => {
            const buttons = [
                { type: 'postback', title: 'Man', payload: 'GENDER_MALE' },
                { type: 'postback', title: 'Woman', payload: 'GENDER_FEMALE' }
            ];
            convo.sendButtonTemplate(`What is your gender?`, buttons);
        }, (payload, convo, data) => {
            const text = payload.message.text;
            convo.set('gender', text);
        }, [
            {
                event: 'postback',
                callback: (payload, convo) => {
                    convo.say('').then(() => askMensage(convo));
                }
            }
        ]);
    };

    const askMensage = (convo) => {
        convo.ask(`Nice, now send for us your message:`, (payload, convo, data) => {
            const text = payload.message.text;
            convo.set('message', text);
            if(text!=undefined)
                convo.say(`Ok, now ask me the answer:`).then(() => askAttachment(convo));
            else
                chat.conversation((convo) => {
                    convo.sendTypingIndicator(1000).then(() => askAttachment(convo));
                });
        });
    };

    const askAttachment = (convo) => {
        convo.ask((convo) => {
            const buttons = [
                { type: 'postback', title: 'Yes', payload: 'SIM' },
                { type: 'postback', title: 'No', payload: 'NAO' }
            ];
            convo.sendButtonTemplate(`Do you need attach something in your message?`, buttons);
        }, (payload, convo, data) => {
            const text = payload.message.text;

        }, [
            {
                event: 'postback:SIM',
                callback: (payload, convo) => {askDataAttachment(convo)}
            },
            {
                event: 'postback:NAO',
                callback: (payload, convo) => {
                    convo.say('Thanks!').then(() => saveDatabase(convo));
                }
            }
        ]);
    };

    const askDataAttachment = (convo) => {
        convo.ask(`Nice,we have all of your credentials. Send for us your attachment:`, (payload, convo, data) => {
            console.log(payload);
            convo.say('Thanks!').then(() => convo.end());
        });

    };

    const saveDatabase = (convo) => {
        convo.ask(`Final question. How old are you?`, (payload, convo, data) => {
            const text = payload.message.text;
            convo.set('age', text);
            convo.say(`That's great!`).then(() => {
                convo.say(`Your information:
      - Campaign_ID ${convo.get('campaign_id')}
      - Name: ${convo.get('name')}
      - email: ${convo.get('email')}
      - Gender: ${convo.get('gender')}
      - Message: ${convo.get('message')}
      `);

                convo.end();
            });
        });
    };


});


bot.start(3000);
```

**Comment by JulienLeal** [comment]
This error appears in this part 
```` // Check for text messages that match a listening pattern
    const patternCallbacks = this.listeningCallbacks.filter(callbackObject => (
      callbackObject.pattern !== undefined
    ));
    if (data.type === 'message' && patternCallbacks.length > 0) {
      for (let i = 0; i < patternCallbacks.length; i += 1) {
        const match = textMatchesPatterns(payload.message.text, patternCallbacks[i].pattern);
        if (match !== false) {
          this.listeningCallbacks = [];
          data.keyword = match.keyword;
          if (match.match) {
            data.match = match.match;
          }
          return patternCallbacks[i].callback.apply(this, [ payload, this, data ]);
        }
      }
    }
```

Do you have any idea charca???

**Comment by kennetham** [comment]
On top of this, how do you call 'quick_replies' location type? I can't seem to pass it in.

Please advise.

**Comment by Charca** [comment]
Hey @JulienLeal, sorry for the huge delay I'm just now finding time to properly debug and fix this one and get the framework in better shape.

So this problem was caused by a race condition when a user sent 2 consecutive messages to the bot before the bot has a chance to ask a question.

When a user sends 2 consecutive messages (with less than 1 second between them), the first message starts the conversation and immediately calls the factory method to ask the first question. But in your example, before asking the first question you send a typing indicator for 1 second, and if during this time the second message comes in, it would be considered an answer (because the conversation has already been started), but it wouldn't know to what question (because the question will be asked after the typing indicator).

So it ended up in this weird state because of this race condition. I've implemented a solution in #15 and I'll merge up and publish a patch if no one has any suggestions.

Thanks for the patience, and let me know if you have any questions.

**Comment by Charca** [comment]
@kennetham I'm not sure I understand your question. Is it related to this issue or just a general question about quick_replies? If so, could you please specify?

**Comment by kennetham** [comment]
@Charca I'm asking in general for quick_replies, is there a way to use the quick_replies with location type etc, seems like its accepting text only at the moment.

**Comment by Charca** [comment]
@kennetham I see what you mean now, I wasn't aware of these new Location and Image quick replies, I created #16 for this and I'll work on it later tonight, should be a pretty quick fix. Thanks for reporting it!

**Comment by Charca** [comment]
Just released v1.0.7 that includes fixes for both issues discussed here cc/ @JulienLeal @kennetham
